### PR TITLE
[Perf] Compute maxsim in worker side, reducing redundant copies, 2.7% E2E throughput improvement

### PR DESCRIPTION
**Pull request by yewentao256**
## Purpose

This PR moves MaxSim scoring from the API server to workers.

This reduces embedding transfer overhead and improves rerank latency/throughput.

Follow up TODO: batched version in worker side.

CC: @noooop @DarkLight1337 

### Acc

Covered in unit tests

```bash
tests/entrypoints/pooling/score/test_online_colbert.py::TestColBERTOnline::test_score
tests/entrypoints/pooling/score/test_online_colbert.py::TestColBERTOnline::test_rerank
tests/entrypoints/pooling/score/test_online_colbert.py::TestColBERTOnline::test_rerank_top_n
...
```

### Perf

`vllm serve   --model jinaai/jina-colbert-v2   --runner pooling   --port 9256   --enforce-eager   --max-model-len 4096   --max-num-batched-tokens 4096   --disable-log-stats   --hf-overrides '{"architectures": ["ColBERTJinaRobertaModel"]}'   --trust-remote-code`

`vllm bench serve   --model jinaai/jina-colbert-v2   --backend vllm-rerank   --endpoint /v1/rerank   --host 127.0.0.1 --port 9256   --dataset-name random-rerank   --num-prompts 2000   --request-rate inf   --max-concurrency 64   --seed 0   --random-input-len 2048   --random-range-ratio 0.5   --percentile-metrics e2el   --metric-percentiles 50,95,99`

```bash
# now
============ Serving Benchmark Result ============
Successful requests:                     2000      
Failed requests:                         0         
Maximum request concurrency:             64        
Benchmark duration (s):                  12.08     
Total input tokens:                      4116113   
Request throughput (req/s):              165.63    
Total token throughput (tok/s):          340878.45 
----------------End-to-end Latency----------------
Mean E2EL (ms):                          380.50    
Median E2EL (ms):                        380.24    
P50 E2EL (ms):                           380.24    
P95 E2EL (ms):                           407.26    
P99 E2EL (ms):                           452.19    
==================================================

# main with gpu
============ Serving Benchmark Result ============
Successful requests:                     2000      
Failed requests:                         0         
Maximum request concurrency:             64        
Benchmark duration (s):                  12.41     
Total input tokens:                      4116113   
Request throughput (req/s):              161.14    
Total token throughput (tok/s):          331637.66 
----------------End-to-end Latency----------------
Mean E2EL (ms):                          391.46    
Median E2EL (ms):                        395.45    
P50 E2EL (ms):                           395.45    
P95 E2EL (ms):                           421.21    
P99 E2EL (ms):                           431.76    
==================================================

# main with cpu
============ Serving Benchmark Result ============
Successful requests:                     2000      
Failed requests:                         0         
Maximum request concurrency:             64        
Benchmark duration (s):                  14.15     
Total input tokens:                      4116113   
Request throughput (req/s):              141.38    
Total token throughput (tok/s):          290961.01 
----------------End-to-end Latency----------------
Mean E2EL (ms):                          446.27    
Median E2EL (ms):                        440.32    
P50 E2EL (ms):                           440.32    
P95 E2EL (ms):                           516.98    
P99 E2EL (ms):                           531.58    
==================================================
```

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -2798,6 +2815,75 @@ def _pool(
...
+
+            if mode == LATE_INTERACTION_MODE_CACHE_QUERY:
+                query_uses = extra_kwargs.get(LATE_INTERACTION_QUERY_USES_KEY, 1)
+                # `output` can be a view into the current step's hidden-states
+                # buffer, so clone it before storing across scheduling steps.
+                self._late_interaction_query_cache[query_key] = output.clone()
+                self._late_interaction_query_uses[query_key] = max(1, int(query_uses))
+                # return a tiny scalar so this query request can finish quickly
+                outputs[i] = torch.zeros((), device=output.device, dtype=torch.float32)
+                continue
```
[the remark this answers is not in the document]
Comment (yewentao256) [reply vllm/v1/worker/gpu_model_runner.py]: Added clear cache. Try expect is not needed

File: vllm/v1/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -482,6 +490,10 @@ def __init__(
 
         # mm_hash ->  encoder_output
         self.encoder_cache: dict[str, torch.Tensor] = {}
+        # query_key -> token embeddings for late-interaction scoring.
+        self._late_interaction_query_cache: dict[str, torch.Tensor] = {}
+        # query_key -> remaining number of docs that should use this query.
+        self._late_interaction_query_uses: dict[str, int] = {}
 
         self.use_aux_hidden_state_outputs = False
```
Comment (noooop) [inline vllm/v1/worker/gpu_model_runner.py:497]: awesome!

My only concern is that this will make the 6,000+ line runner code even longer, and I don’t know if the core members will be happy with it.

Comment (DarkLight1337) [reply vllm/v1/worker/gpu_model_runner.py:497]: I agree this will not fly with core maintainers. Maybe it'll be more feasible to implement this in a modular way under model runner v2

Comment (yewentao256) [reply vllm/v1/worker/gpu_model_runner.py:497]: @noooop @DarkLight1337 Nice catch, I have refactored the file so that we don't involve too much in model runner. I don't want to involve V2 in this PR since there are lots of PRs going on for V2 and woosuk is too busy to review more PRs, which might block us for weeks.

File: vllm/v1/engine/core_client.py
Comment (njhill) [inline vllm/v1/engine/core_client.py:1]: We should similarly avoid these feature-specific changes to core_client.py. `_get_late_interaction_engine_index` shouldn't live in this file

Comment (yewentao256) [reply vllm/v1/engine/core_client.py:1]: Thanks! Refactored `_get_late_interaction_engine_index` out of this file

File: vllm/v1/pool/late_interaction.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+
+def get_late_interaction_engine_index(
+    pooling_params: PoolingParams | None,
+    num_engines: int,
+) -> int | None:
+    if pooling_params is None or pooling_params.extra_kwargs is None:
+        return None
+
+    extra_kwargs = pooling_params.extra_kwargs
```
Comment (DarkLight1337) [inline vllm/v1/pool/late_interaction.py]: I feel that we should just add a `LateInteractionParams` as a sub-config to `PoolingParams`, as accessing hard-coded keys in something as generic as `extra_kwargs` feels a bit hacky.

Comment (yewentao256) [reply vllm/v1/pool/late_interaction.py]: Nice catch, fixed

File: vllm/pooling_params.py [hunk 4]
```diff
@@ -11,6 +11,16 @@
 from vllm.tasks import PoolingTask
 
 
+class LateInteractionParams(
+    msgspec.Struct,
+    omit_defaults=True,  # type: ignore[call-arg]
+    array_like=True,
+):  # type: ignore[call-arg]
+    mode: str
```
Comment (DarkLight1337) [inline vllm/pooling_params.py:29]: Can you add docstrings?

Comment (yewentao256) [reply vllm/pooling_params.py:29]: Added~

File: vllm/v1/worker/gpu/pool/late_interaction_runner.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+class LateInteractionRunner:
+    """Worker-side state and postprocessing for late-interaction scoring."""
+
+    def __init__(self) -> None:
+        # query_key -> token embeddings for late-interaction scoring.
+        self._query_cache: dict[str, torch.Tensor] = {}
+        # query_key -> remaining number of docs that should use this query.
+        self._query_uses: dict[str, int] = {}
+        # doc request id -> query key.
+        self._doc_query_keys: dict[str, str] = {}
```
Comment (noooop) [inline vllm/v1/worker/gpu/pool/late_interaction_runner.py:25]: Is setting up a dedicated Runner for LateInteraction and LateInteractionParams overkill?

Comment (noooop) [reply vllm/v1/worker/gpu/pool/late_interaction_runner.py:25]: Only for 2.7% E2E throughput improvement

Comment (yewentao256) [reply vllm/v1/worker/gpu/pool/late_interaction_runner.py:25]: This is the naive scalar version, we haven't really optimized the performance yet.
We can do much better if we have bmm/specific kernel for it, this PR is just for the base.

File: vllm/v1/worker/gpu/pool/late_interaction_runner.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        if mode == LATE_INTERACTION_MODE_SCORE_DOC and query_key is not None:
+            self._doc_query_keys[req_id] = query_key
+        else:
+            self._doc_query_keys.pop(req_id, None)
+
+    def on_requests_finished(self, finished_req_ids: Iterable[str]) -> None:
+        for req_id in finished_req_ids:
+            query_key = self._doc_query_keys.pop(req_id, None)
+            if query_key is not None:
+                self._release_query_use(query_key)
```
[the remark this answers is not in the document]
Comment (yewentao256) [reply vllm/v1/worker/gpu/pool/late_interaction_runner.py:45]: We won't reach there since we have `_validate_score_input_lens`

**Review by DarkLight1337 — APPROVED** [review]
